### DISCUSSION TITLE
Is there a way to partially influence the parametrization of a type variable from an `__init__`/`__new__` argument?

### DISCUSSION THREAD

**Discussion by Viicos** [microsoft/pyright#6782] [Q&A]
I'm trying to find a solution to type check django fields, in a way that doesn't rely on the current plugin. The signature of `Field` contains a `null: bool` argument, which when set to `True`, allow `None` as a model instance value for this field.

The current implementation of `django-stubs` is to add types to the `__set__`/`__get__` methods of the `Field` class. Nullability is currently handled in the plugin. I've been able to implement the following (title says _partially_ because I'm just adding `| None` to the type variables):

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.13&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDiApikSEgMYA0UAKokQGoCGINAMkjGcwDY1gAbmV5hmAEwBQ0gPoBlWlAC8dBixAAKAETzaWgJSSZBRSvoImrbcb2HpFXswDOTqADEkRXuI3FS5CgBtXRobAF19AC5JKFioAAEhETEpOKhxImAoGRlSAHccjRi0uIcnKmKSqAAqCqrYlABXXl5IqE5uED5A2hBGojC6uP0oAFoAPndPb2CFKAAfKAA5MFJQkwXl1YG2gDp9yoSkkFEJQ4ysnPzCw7Syoara27imlraOnl5Atz4nAeUoPtdg8oCMJlMvOJZrR1rQwnsDocAMTuXDpIgQVZOGBdGBIVZQBCNEA4P5OaJpC5QNBEGAyeAWDR-XjAMGTGwI3bnTJQP50hlEJleVljdkKTn2RwuKAASRQ3BpIA8kI0ypmwFO%2BEWqC1vJxNB1EQpcQQzic0kkzABcoVZDVPlevCUvX6hhARGEfHpDA0zF2NP5Pv0IygKPMRCgYCyWj9AY0%2Bi0mFcWh1idiAGotqRMFlHc6%2BkRJO7PbxvYy-Xyy4Lg7EwwxI9GK7T44mkMmNWJdTrNtiQInMyts0hc80nS7C0A)

```python
from typing import Generic, TypeVar, Literal, overload


_ST = TypeVar("_ST")
_GT = TypeVar("_GT")


class Field(Generic[_ST, _GT]):
    @overload
    def __new__(
        cls,
        *,
        null: Literal[True],
    ) -> Field[_ST | None, _GT | None]: ...
    @overload
    def __new__(
        cls,
        *,
        null: Literal[False] = ...,
    ) -> Field[_ST, _GT]: ...

    # For demonstration purposes:
    def get_type(self) -> _GT: ...
    def set_type(self) -> _ST: ...

class IntegerField(Field[float | int | str, int]):
    pass


a = IntegerField(null=True)
reveal_type(a.get_type())  # Type of "a.get()" is "int"   + None if null=True
reveal_type(a.set_type())  # Type of "a.set()" is "float | int | str" + None if null=True
```

Unfortunately, the return type of the `__new__` method is incorrect:

```python
reveal_type(a)  # Type of "a" is "Field[float | int | str | None, int | None]", should be `IntegerField`
```

I could use `Self` as a return type instead, but `Self` [can't be used with type arguments](https://peps.python.org/pep-0673/#use-in-generic-classes) (so I can't use `-> Self[_ST, _GT]`).

Considering this, I've tried going the old way (used before `Self` was a thing) ([playground](https://pyright-play.net/?pythonVersion=3.13&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFAOICmKlISAxmQCqKUBqAhiGQDJIy0OAGzIBlSkOBkwAN1pCwHACYAoNQH1RTKAF4oLBOy4AKAESampgJQr15bXoNGQZu5ZsqAYkglKH%2B1k4XU29fSzIAIzAAVxQlHRCfISVrNRV6IQ4AZyyoUOTjKho6egBtCzI3AF0rAC4VKEaoAAFZeUVVJqglSmAodXUaAHcB4waupoys2thWUvy-cq1K%2ByqqknGJqAAqDa3GlGihIRm%2BARBhUqYQaMp1zagrKABaAD48pMWLKAAfKAA5MA0FbaP6AmhVGYAOhhD1achACmUDx6fQGw1GDy6Uxm8EM80%2BTCWTBBaz2%2B12WKah2Op34giE82EWTuuigMKh5MaTzeHzCxNJ0NhDwAxHlcN1KBAgVkYBcYEggVAENEQDgWdMUb0oGhKDB1HjKMYWZIee83EKoVq%2Biz9YbjRJgGb%2BlpLWoMtlcgBJFACXUgBbGBalYBI-B-VDhqCy7iYX01epdBCetIcNk%2Bv20QM0oQ6a63GwgShyYQG1jGDhQ3V28tWJ5QMVOKBgPqmSvV4xWUyYXKmSOmMgQJA5VAYAAG4MoY8wfRzeZulCgtQA9CoiyWhGXDBWobat0a641G6xm63K7bO93h1BTKHFFHI79o3KB1AhyP0FAJ0CpzOoHP80XFc12LShS3tDgrCAA)), but without success.

Using `__init__(self: Field[..., ...]` instead of hacking with `__new__` leads to the same result, i.e. nullability ins't taken into account.

I found this to be related in some way: https://github.com/python/typing/discussions/1197, although it differs a bit from what I have. Seems like I'm also hitting the need for higher kinded types.

I'm wondering if there's a solution to this. Thanks in advance!

Edit: to be clear, this can still be done by explicitly writing the `__new__` definitions on each subclass (see [this issue](https://github.com/microsoft/pyright/issues/1211), which is closely related and lead to a working solution using `__init__`). What I am asking here is if there's a way to make it possible to have the same behavior without these explicit overloads on each subclass.

**Comment by erictraut** [comment]
@Viicos, I see you closed this discussion in favor of a [python/typing discussion](https://discuss.python.org/t/is-there-a-way-to-partially-influence-the-parametrization-of-a-type-variable-from-an-init-new-argument/41483). That's a good idea since your question will get more exposure from the typing community. I'll spend some time thinking about your question above and respond to the other thread if I come up with a good answer.

BTW, I read your [recent article](https://viicos.github.io/posts/an-alternative-to-the-django-mypy-plugin/). Great analysis. I'd love to see static typing and django work together better, and I'm glad to see you're making this investment. Let me know if there are ways I can help support your effort.

**Reply by Viicos** [reply]
Figured out that the question wasn't particularly tied to `pyright` indeed.

> I'd love to see static typing and django work together better

Hopefully this would be a reliable solution. My main concern is performance, LibCST start to struggle on semi-large projects. As always, thanks for the quick feedback on questions and issues!
